src/main/java/pl/termosteam/kinex/controller/ClientTicketController.java

```diff
@@ -0,0 +1,79 @@
+import org.modelmapper.ModelMapper;
+import org.springframework.http.ResponseEntity;
+import org.springframework.security.access.prepost.PreAuthorize;
+import org.springframework.web.bind.annotation.*;
+import pl.termosteam.kinex.domain.Role;
+import pl.termosteam.kinex.domain.Ticket;
+import pl.termosteam.kinex.domain.User;
+import pl.termosteam.kinex.dto.TicketCancelRequestDto;
+import pl.termosteam.kinex.dto.TicketRequestAdminDto;
+import pl.termosteam.kinex.dto.TicketRequestClientDto;
+import pl.termosteam.kinex.dto.TicketResponseDto;
+import pl.termosteam.kinex.service.TicketService;
+import pl.termosteam.kinex.service.UserService;
+
+import javax.validation.Valid;
+import java.util.List;
+
+@AllArgsConstructor
+@RestController
+@RequestMapping("api/client/ticket")
+public class ClientTicketController {
+
+    private final TicketService ticketService;
+    private final ModelMapper mm;
+    private final UserService userService;
+
+    @PreAuthorize("hasRole('ROLE_USER')")
+    @PostMapping(path = "make-reservation")
+    public TicketResponseDto[] makeReservation(@RequestBody @Valid TicketRequestClientDto ticketInfo) {
+        User reservedByUser = userService.getUserNotNullIfAuthenticated();
+
+        TicketRequestAdminDto ticketFullInfo = new TicketRequestAdminDto(
+                reservedByUser.getId(),
+                ticketInfo.getScreeningId(),
+                ticketInfo.getSeatIds());
+
+        List<Ticket> tickets = ticketService.makeReservation(ticketFullInfo, reservedByUser, Role.USER);
+
+        return mm.map(tickets, TicketResponseDto[].class);
+    }
+
+    @PreAuthorize("hasRole('ROLE_USER')")
+    @PutMapping(path = "cancel-reservation/{screeningId}")
+    public String cancelReservationForScreening(@PathVariable int screeningId) {
+        int userId = userService.getUserNotNullIfAuthenticated().getId();
+
+        TicketCancelRequestDto ticketInfo = new TicketCancelRequestDto(userId, screeningId);
+
+        return ticketService.cancelReservationForScreening(ticketInfo);
+    }
+
+    @PreAuthorize("hasRole('ROLE_USER')")
+    @GetMapping
+    public TicketResponseDto[] findUserTickets() {
+        int userId = userService.getUserNotNullIfAuthenticated().getId();
+
+        List<Ticket> tickets = ticketService.findUserTickets(userId);
+
+        return mm.map(tickets, TicketResponseDto[].class);
+    }
+
+    @PreAuthorize("hasRole('ROLE_USER')")
+    @GetMapping(path = "screening/{screeningId}")
+    public ResponseEntity<?> findUserTicketsForScreening(@PathVariable int screeningId) {
+        int userId = userService.getUserNotNullIfAuthenticated().getId();
+
+        List<Ticket> tickets = ticketService.findUserTicketsForScreening(userId, screeningId);
+
+        if (CollectionUtils.isNotEmpty(tickets)) {
+            return ResponseEntity.ok(mm.map(tickets, TicketResponseDto[].class));
+        } else {
+            return ResponseEntity.ok("You have not purchased any tickets for this screening.");
```

Comment:
Starajcie się unikać sytuacji gdy endpoint zwraca typ według uznania (raz tablica, raz string). Znacznie utrudnia to potem obsługę tego endpointu po stronie klienta.

W takiej sytuacji możecie zwrócić pustą tablicę lub kod 404.

Reply:
ok, poprawilem na pusta tablice dd967b84